### PR TITLE
Add comprehensive input validation with detailed error responses

### PR DESCRIPTION
**Pull request by olaservo**
## Summary
- Added comprehensive input validation middleware for all Todo endpoints
- Implemented field-specific error messages with structured error responses  
- Added validation for title length, HTML tag prevention, and type checking
- Enhanced ID parameter validation for route security
- Added comprehensive test coverage for all validation scenarios

### New Files
- `src/middleware/validation.ts` - Core validation logic with custom error handling
- `src/middleware/validation.test.ts` - Comprehensive test suite (15+ test cases)

### Modified Files
- `src/routes/todos.ts` - Integrated validation middleware, improved error responses
- `README.md` - Added validation rules and error response format documentation
- `package.json` - Added supertest for API testing

### Validation Rules Implemented
- **Title**: Required, 1-100 chars, no HTML tags, string type validation
- **Description**: Optional, max 500 chars, no HTML tags, string type validation  
- **Completed**: Boolean type validation when provided
- **ID Parameters**: Positive integer validation for all routes

## Test Coverage
- Valid input acceptance
- Missing required fields
- Invalid data types
- Length limit enforcement
- HTML tag prevention (XSS protection)
- Multiple simultaneous errors
- Edge cases and boundary conditions

## API Response Changes
**Before**: Simple error strings
```json
{ "error": "Title is required" }
```

**After**: Structured error objects with field details
```json
{
  "error": "Validation failed", 
  "details": [
    { "field": "title", "message": "Title is required" },
    { "field": "description", "message": "Description must be 500 characters or less" }
  ]
}
```

## Security Improvements
- HTML tag prevention to mitigate XSS attacks
- Input sanitization with automatic trimming
- Type validation to prevent injection attacks
- ID format validation to prevent path traversal

## Testing Instructions

1. **Install dependencies**: `npm install`
2. **Run test suite**: `npm test` (should show 15+ passing tests)
3. **Test API manually**:
   ```bash
   npm run dev
   
   # Test valid creation
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"Valid Todo","description":"Valid description"}'
   
   # Test validation errors
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"","description":"'$(printf 'a%.0s' {1..501})'"}'
   ```

## Breaking Changes
**Error response format has changed** - API consumers will need to handle the new structured error format

## Future Enhancements
- [ ] Add rate limiting validation
- [ ] Implement custom validation rules configuration
- [ ] Add internationalization for error messages
- [ ] Consider adding request sanitization middleware

### CODE REVIEW

File: src/routes/todos.ts [hunk 1]
```diff
@@ -9,54 +10,67 @@ router.get('/', (req: Request, res: Response) => {
...
+router.post('/', validateTodoCreate, (req: Request, res: Response) => {
   const { title, description }: CreateTodoRequest = req.body;
   
-  if (!title || title.trim() === '') {
-    return res.status(400).json({ error: 'Title is required' });
-  }
-  
-  const todo = todoStore.create({ title: title.trim(), description });
+  // BUG: No try-catch block - if todoStore.create throws, server will crash
+  const todo = todoStore.create({
```
Comment (dandeliongold) [inline src/routes/todos.ts:31]: **🐛 Potential Crash Risk: Missing Error Handling**

If `todoStore.create()` throws an exception, the server will crash because there's no try-catch block around this operation.

**Recommendation**: Add error handling:
```typescript
try {
  const todo = todoStore.create({ 
    title: title.trim(), 
    description: description?.trim() || undefined 
  });
  res.status(201).json(todo);
} catch (error) {
  res.status(500).json({ 
    error: 'Failed to create todo', 
    details: [{ field: 'server', message: 'Internal server error' }] 
  });
}
```

File: src/middleware/validation.test.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        .post('/')
+        .send({ title: '<script>alert("xss")</script>' });
+
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'title',
+        message: 'Title cannot contain HTML tags'
+      });
+    });
+
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:61]: **📝 Missing Test Coverage**

The comment indicates missing test coverage for description HTML validation. This is important for security (XSS prevention).

**Recommendation**: Add this test case:
```typescript
it('should reject description with HTML tags', async () => {
  const response = await request(app)
    .post('/')
    .send({ title: 'Valid Title', description: '<script>alert("xss")</script>' });

  expect(response.status).toBe(400);
  expect(response.body.details).toContainEqual({
    field: 'description',
    message: 'Description cannot contain HTML tags'
  });
});
```

File: src/middleware/validation.test.ts [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+      const response = await request(app).get('/abc');
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'id',
+        message: 'ID must be a positive integer',
+        value: 'abc'
+      });
+    });
+
+    // MISSING TEST: Should test ID "0" edge case - currently no validation for this
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:120]: **📝 Missing Edge Case Test**

The comment indicates missing test for ID "0" edge case, which aligns with the validation bug I identified.

**Recommendation**: Add this test:
```typescript
it('should reject ID "0"', async () => {
  const response = await request(app).get('/0');
  expect(response.status).toBe(400);
  expect(response.body.details).toContainEqual({
    field: 'id',
    message: 'ID must be a positive integer',
    value: '0'
  });
});
```

File: src/middleware/validation.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
+import { Request, Response, NextFunction } from 'express';
+
+export interface ValidationError {
+  field: string;
+  message: string;
+  value?: any;
+}
+
+export class ValidationException extends Error {
```
Comment (dandeliongold) [inline src/middleware/validation.ts:9]: **💡 Improvement: Unused ValidationException Class**

The `ValidationException` class is defined but never used in the codebase. The validation functions return response objects directly instead of throwing exceptions.

**Options:**
1. Remove the unused class to reduce code clutter
2. Or refactor to use it consistently:
```typescript
// Instead of returning response directly:
if (errors.length > 0) {
  throw new ValidationException(errors);
}
// Handle in middleware or error handler
```
